File: innerlibs1/ConsoleExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Console

    ''' <summary>
    ''' Métodos para manipulação de aplicações baseadas em Console (System.Console)
    ''' </summary>
    Public Module ConsoleExtensions

        ''' <summary>
        ''' Escreve no console colorindo palavras especificas
        ''' </summary>
        ''' <param name="Text">Texto</param>
        ''' <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>
        <Extension()> Public Function ConsoleWrite(Text As String, CustomColoredWords As Dictionary(Of String, ConsoleColor), Optional Lines As Integer = 0) As String
            Return ConsoleWrite(Text, CustomColoredWords, StringComparison.InvariantCultureIgnoreCase, Lines)
        End Function

        ''' <summary>
        ''' Escreve no console colorindo palavras especificas
        ''' </summary>
        ''' <param name="Text">Texto</param>
        ''' <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>
        ''' <param name="Comparison">Tipo de comparação</param>
        <Extension()> Public Function ConsoleWrite(Text As String, CustomColoredWords As Dictionary(Of String, ConsoleColor), Comparison As StringComparison, Optional BreakLines As Integer = 0) As String
            CustomColoredWords = If(CustomColoredWords, New Dictionary(Of String, ConsoleColor))

            CustomColoredWords = CustomColoredWords.SelectMany(Function(x) x.Key.Split(" ").Distinct().ToDictionary(Function(y) y, Function(y) x.Value)).ToDictionary()

            Dim lastcolor = System.Console.ForegroundColor
            Dim maincolor = CustomColoredWords.GetValueOr("", lastcolor)
            If Text.IsNotBlank Then
                Dim substrings As String() = Text.Split(" ")
                For Each substring As String In substrings
                    System.Console.ForegroundColor = maincolor
                    System.Console.ForegroundColor = CustomColoredWords.Where(Function(x) x.Key.Equals(substring, Comparison)).Select(Function(x) x.Value).FirstOr(maincolor)
                    System.Console.Write(substring & " ")
                Next
            End If
            System.Console.ForegroundColor = lastcolor
            ConsoleBreakLine(BreakLines)
            Return Text
        End Function

        ''' <summary>
        ''' Escreve no console usando uma cor especifica
        ''' </summary>
        ''' <param name="Text">Texto</param>
        ''' <param name="Color">Cor</param>
        <Extension()> Public Function ConsoleWrite(Text As String, Optional Color As ConsoleColor = ConsoleColor.White, Optional BreakLines As Integer = 0) As String
            Dim lastcolor = System.Console.ForegroundColor
            System.Console.ForegroundColor = Color
            System.Console.Write(Text)
            System.Console.ForegroundColor = lastcolor
            ConsoleBreakLine(BreakLines)
            Return Text
        End Function

        ''' <summary>
        ''' Escreve uma linha no console colorindo palavras especificas
        ''' </summary>
        ''' <param name="Text">Texto</param>
        ''' <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>

        <Extension()> Public Function ConsoleWriteLine(Text As String, CustomColoredWords As Dictionary(Of String, ConsoleColor), Optional Lines As Integer = 1) As String
            Lines = Lines.SetMinValue(1)
            ConsoleWrite(Text, CustomColoredWords, Lines)
            Return Text
        End Function

        ''' <summary>
        ''' Escreve uma linha no console usando uma cor especifica
        ''' </summary>
        ''' <param name="Text">Texto</param>
        ''' <param name="Color">Cor</param>

        <Extension()> Public Function ConsoleWriteLine(Text As String, Optional Color As ConsoleColor = ConsoleColor.White, Optional Lines As Integer = 1) As String
            Lines = Lines.SetMinValue(1)
            ConsoleWrite(Text, Color, Lines)
            Return Text
        End Function


        ''' <summary>
        ''' Escreve o texto de uma exception no console
        ''' </summary>
        ''' <param name="Exception">Texto</param>
        ''' <param name="Color">Cor</param>

        <Extension()> Public Function ConsoleWriteError(Of T As Exception)(Exception As T, Optional Color As ConsoleColor = ConsoleColor.Red, Optional Lines As Integer = 1) As T
            Lines = Lines.SetMinValue(1)
            ConsoleWrite(Exception.ToFullExceptionString, Color, Lines)
            Return Exception
        End Function

        ''' <summary>
        ''' Pula uma ou mais linhas no console
        ''' </summary>
        ''' <param name="Lines">Numero de linhas</param>
        Public Sub ConsoleBreakLine(Optional Lines As Integer = 1)
            If Lines > 0 Then
                For index = 1 To Lines
                    System.Console.WriteLine("")
                Next
            End If
        End Sub

        ''' <summary>
        ''' Pula uma ou mais linhas no console e retorna a mesma string (usada como chaining)
        ''' </summary>
        ''' <param name="Lines">Numero de linhas</param>
        <Extension()> Public Function ConsoleBreakLine(Text As String, Optional Lines As Integer = 1) As String
            ConsoleBreakLine(Lines)
            Return Text
        End Function

        ''' <summary>
        ''' Le a proxima linha inserida no console pelo usuário
        ''' </summary>
        ''' <returns></returns>
        Public Function ReadLine() As String
            Return System.Console.ReadLine
        End Function

        ''' <summary>
        ''' Le o proximo caractere inserido no console pelo usuário
        ''' </summary>
        ''' <returns></returns>
        Public Function ReadChar() As Char
            Return System.Console.ReadKey.KeyChar
        End Function


        ''' <summary>
        ''' Le a proxima tecla pressionada pelo usuário
        ''' </summary>
        ''' <returns></returns>
        Public Function ReadKey() As ConsoleKey
            Return System.Console.ReadKey.Key
        End Function

        ''' <summary>
        ''' Toca um Beep
        ''' </summary>
        ''' <param name="Times">Numero de beeps</param>
        Public Sub Beep(Optional Times As Integer = 1)
            For index = 1 To Times.SetMinValue(1)
                System.Console.Beep()
            Next
        End Sub

        ''' <summary>
        ''' Toca um beep especifico
        ''' </summary>
        ''' <param name="Frequency">Frequencia</param>
        ''' <param name="Duration">Duracao em milisegundos</param>
        Public Sub Beep(Frequency As Integer, Duration As Integer)
            System.Console.Beep(Frequency.LimitRange(37, 32767), Duration)
        End Sub

        ''' <summary>
        ''' Titulo da janela do console
        ''' </summary>
        ''' <returns></returns>
        Public Property Title As String
            Get
                Return System.Console.Title
            End Get
            Set(value As String)
                System.Console.Title = value
            End Set
        End Property

    End Module

End Namespace

File: InnerLibs/ConsoleExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace InnerLibs.Console
{

    /// <summary>
    /// Métodos para manipulação de aplicações baseadas em Console (System.Console)
    /// </summary>
    public static class ConsoleExtensions
    {

        /// <summary>
        /// Escreve uma data com descrição no console
        /// </summary>
        /// <param name="dateTime"></param>
        /// <param name="Text"></param>
        /// <param name="BreakLines"></param>
        /// <returns></returns>
        public static string ConsoleWrite(this DateTime dateTime, string Text,int BreakLines = 0) => ConsoleWrite($"{dateTime} - {Text}",BreakLines);

        /// <summary>
        /// Escreve uma data com descrição no console e quebra 1 ou mais linhas
        /// </summary>
        /// <param name="dateTime"></param>
        /// <param name="Text"></param>
        /// <param name="BreakLines"></param>
        /// <returns></returns>
        public static string ConsoleWriteLine(this DateTime dateTime, string Text,int BreakLines = 1) => dateTime.ConsoleWrite(Text,BreakLines.SetMinValue(1));

        /// <summary>
        /// Escreve no console colorindo palavras especificas
        /// </summary>
        /// <param name="Text">Texto</param>
        /// <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>
        public static string ConsoleWrite(this string Text, Dictionary<string, ConsoleColor> CustomColoredWords, int Lines = 0) => Text.ConsoleWrite(CustomColoredWords, StringComparison.InvariantCultureIgnoreCase, Lines);

        /// <summary>
        /// Escreve no console colorindo palavras especificas
        /// </summary>
        /// <param name="Text">Texto</param>
        /// <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>
        /// <param name="Comparison">Tipo de comparação</param>
        public static string ConsoleWrite(this string Text, Dictionary<string, ConsoleColor> CustomColoredWords, StringComparison Comparison, int BreakLines = 0)
        {
            CustomColoredWords = CustomColoredWords ?? new Dictionary<string, ConsoleColor>();
            CustomColoredWords = CustomColoredWords.SelectMany(x => x.Key.Split(" ").Distinct().ToDictionary(y => y, y => x.Value)).ToDictionary();
            var lastcolor = System.Console.ForegroundColor;
            var maincolor = CustomColoredWords.GetValueOr("", lastcolor);
            if (Text.IsNotBlank())
            {
                var substrings = Text.Split(" ");
                foreach (string substring in substrings)
                {
                    System.Console.ForegroundColor = maincolor;
                    System.Console.ForegroundColor = CustomColoredWords.Where(x => x.Key.Equals(substring, Comparison)).Select(x => x.Value).FirstOr(maincolor);
                    System.Console.Write(substring + " ");
                }
            }

            System.Console.ForegroundColor = lastcolor;
            ConsoleBreakLine(BreakLines);
            return Text;
        }

        /// <summary>
        /// Escreve no console usando uma cor especifica
        /// </summary>
        /// <param name="Text">Texto</param>
        public static string ConsoleWrite(this string Text, int BreakLines = 0) => Text.ConsoleWrite(System.Console.ForegroundColor, BreakLines);

        /// <summary>
        /// Escreve no console usando uma cor especifica
        /// </summary>
        /// <param name="Text">Texto</param>
        /// <param name="Color">Cor</param>
        public static string ConsoleWrite(this string Text, ConsoleColor Color, int BreakLines = 0)
        {
            var lastcolor = System.Console.ForegroundColor;
            System.Console.ForegroundColor = Color;
            System.Console.Write(Text ?? "");
            System.Console.ForegroundColor = lastcolor;
            ConsoleBreakLine(BreakLines);
            return Text;
        }

        /// <summary>
        /// Escreve uma linha no console colorindo palavras especificas
        /// </summary>
        /// <param name="Text">Texto</param>
        /// <param name="CustomColoredWords">Lista com as palavras e suas respectivas cores</param>

        public static string ConsoleWriteLine(this string Text, Dictionary<string, ConsoleColor> CustomColoredWords, int Lines = 1)
        {
            Lines = Lines.SetMinValue(1);
            Text.ConsoleWrite(CustomColoredWords, Lines);
            return Text;
        }

        /// <summary>
        /// Escreve uma linha no console usando uma cor especifica
        /// </summary>
        /// <param name="Text">Texto</param>
        /// <param name="Color">Cor</param>

        public static string ConsoleWriteLine(this string Text, ConsoleColor Color, int Lines = 1)
        {
            Lines = Lines.SetMinValue(1);
            Text.ConsoleWrite(Color, Lines);
            return Text;
        }

        /// <summary>
        /// Escreve uma linha no console usando uma cor especifica
        /// </summary>
        /// <param name="Text">Texto</param>
        public static string ConsoleWriteLine(this string Text, int Lines = 1) => Text.ConsoleWriteLine(System.Console.ForegroundColor, Lines);

        /// <summary>
        /// Escreve o texto de uma exception no console
        /// </summary>
        /// <param name="Exception">Texto</param>
        /// <param name="Message">Mensagem Adicional de erro</param>
        /// <param name="Color">Cor</param>
        public static T ConsoleWriteError<T>(this T Exception, string Message, string Separator, ConsoleColor Color = ConsoleColor.Red, int Lines = 1) where T : Exception
        {
            var ex = new Exception(Message.IfBlank("Error"), Exception);
            return (T)ex.ConsoleWriteError(Separator, Color, Lines);
        }

        /// <summary>
        /// Escreve o texto de uma exception no console
        /// </summary>
        /// <param name="Exception">Texto</param>
        /// <param name="Color">Cor</param>
        public static T ConsoleWriteError<T>(this T Exception, string Separator, ConsoleColor Color = ConsoleColor.Red, int Lines = 1) where T : Exception
        {
            Lines = Lines.SetMinValue(1);
            Exception.ToFullExceptionString(Separator).ConsoleWrite(Color, Lines);
            return Exception;
        }

        /// <summary>
        /// Escreve o texto de uma exception no console
        /// </summary>
        public static T ConsoleWriteError<T>(this T Exception, string Message) where T : Exception
        {
            return Exception.ConsoleWriteError(Message, " >> ");
        }

        /// <summary>
        /// Escreve o texto de uma exception no console
        /// </summary>
        public static T ConsoleWriteError<T>(this T Exception) where T : Exception
        {
            return Exception.ConsoleWriteError(" >> ", ConsoleColor.Red, 1);
        }

        public static T ConsoleWriteError<T>(this T Exception, ConsoleColor Color, int Lines = 1) where T : Exception
        {
            return Exception.ConsoleWriteError(" >> ", Color, Lines);
        }

        public static T ConsoleWriteError<T>(this T Exception, int Lines) where T : Exception
        {
            return Exception.ConsoleWriteError(" >> ", ConsoleColor.Red, Lines);
        }

        /// <summary>
        /// Pula uma ou mais linhas no console
        /// </summary>
        /// <param name="Lines">Numero de linhas</param>
        public static void ConsoleBreakLine(int Lines = 1)
        {
            while (Lines > 0)
            {
                System.Console.WriteLine(string.Empty);
                Lines = Lines - 1;
            }
        }

        /// <summary>
        /// Pula uma ou mais linhas no console e retorna a mesma string (usada como chaining)
        /// </summary>
        /// <param name="Lines">Numero de linhas</param>
        public static string ConsoleBreakLine(this string Text, int Lines = 1)
        {
            ConsoleBreakLine(Lines);
            return Text;
        }

        /// <summary>
        /// Le a proxima linha inserida no console pelo usuário
        /// </summary>
        /// <returns></returns>
        public static string ReadLine()
        {
            return System.Console.ReadLine();
        }

        /// <summary>
        /// Le o proximo caractere inserido no console pelo usuário
        /// </summary>
        /// <returns></returns>
        public static char ReadChar()
        {
            return System.Console.ReadKey().KeyChar;
        }

        /// <summary>
        /// Le a proxima tecla pressionada pelo usuário
        /// </summary>
        /// <returns></returns>
        public static ConsoleKey ReadKey()
        {
            return System.Console.ReadKey().Key;
        }

        /// <summary>
        /// Toca um Beep
        /// </summary>
        /// <param name="Times">Numero de beeps</param>
        public static void Beep(int Times = 1)
        {
            for (int index = 1, loopTo = Times.SetMinValue(1); index <= loopTo; index++)
                System.Console.Beep();
        }

        /// <summary>
        /// Toca um beep especifico
        /// </summary>
        /// <param name="Frequency">Frequencia</param>
        /// <param name="Duration">Duracao em milisegundos</param>
        public static void Beep(int Frequency, int Duration, int Times = 1)
        {
            for (int index = 1, loopTo = Times.SetMinValue(1); index <= loopTo; index++)
                System.Console.Beep(Frequency.LimitRange(37, 32767), Duration);
        }
    }
}